libnm/nm-device-wireguard.c

```diff
@@ -0,0 +1,254 @@
+
+G_DEFINE_TYPE (NMDeviceWireguard, nm_device_wireguard, NM_TYPE_DEVICE)
+
+#define NM_DEVICE_WIREGUARD_GET_PRIVATE(o) (G_TYPE_INSTANCE_GET_PRIVATE ((o), NM_TYPE_DEVICE_WIREGUARD, NMDeviceWireguardPrivate))
+
+typedef struct {
+	guchar *private_key;
+	guchar *public_key;
+	guint16 listen_port;
+	guint32 fwmark;
+} NMDeviceWireguardPrivate;
+
+NM_GOBJECT_PROPERTIES_DEFINE_BASE (
+	PROP_PRIVATE_KEY,
+	PROP_PUBLIC_KEY,
+	PROP_LISTEN_PORT,
+	PROP_FWMARK,
+);
+
+/**
+ * nm_device_wireguard_get_private_key:
+ * @device: a #NMDeviceWireguard
+ *
+ * Gets the private key for this interface
+ *
+ * Returns: pointer to the 32-byte private key
+ *
+ * Since: 1.14
+ **/
+const guchar*
+nm_device_wireguard_get_private_key (NMDeviceWireguard *device)
+{
+	g_return_val_if_fail (NM_IS_DEVICE_WIREGUARD (device), 0);
+
+	return NM_DEVICE_WIREGUARD_GET_PRIVATE (device)->private_key;
+}
+
+/**
+ * nm_device_wireguard_get_public_key:
+ * @device: a #NMDeviceWireguard
+ *
+ * Gets the public key for this interface
+ *
+ * Returns: pointer to the 32-byte public key
+ *
+ * Since: 1.14
+ **/
+const guchar*
+nm_device_wireguard_get_public_key (NMDeviceWireguard *device)
+{
+	g_return_val_if_fail (NM_IS_DEVICE_WIREGUARD (device), 0);
+
+	return NM_DEVICE_WIREGUARD_GET_PRIVATE (device)->public_key;
+}
+
+/**
+ * nm_device_wireguard_get_listen_port:
+ * @device: a #NMDeviceWireguard
+ *
+ * Gets the local UDP port this interface listens on
+ *
+ * Returns: UDP listen port
+ *
+ * Since: 1.14
+ **/
+guint
+nm_device_wireguard_get_listen_port (NMDeviceWireguard *device)
+{
+	g_return_val_if_fail (NM_IS_DEVICE_WIREGUARD (device), 0);
+
+	return NM_DEVICE_WIREGUARD_GET_PRIVATE (device)->listen_port;
+}
+
+/**
+ * nm_device_wireguard_get_fwmark:
+ * @device: a #NMDeviceWireguard
+ *
+ * Gets the firewall mark (fwmark) byte, if this interface sets one
+ *
+ * Returns: 0 if fwmark not in use, fwmark byte value otherwise
+ *
+ * Since: 1.14
+ **/
+guint
+nm_device_wireguard_get_fwmark (NMDeviceWireguard *device)
+{
+	g_return_val_if_fail (NM_IS_DEVICE_WIREGUARD (device), 0);
+
+	return NM_DEVICE_WIREGUARD_GET_PRIVATE (device)->fwmark;
+}
+
+/***********************************************************/
+
+static void
+nm_device_wireguard_init (NMDeviceWireguard *device)
+{
+}
+
+static void
+init_dbus (NMObject *object)
+{
+	NMDeviceWireguardPrivate *priv = NM_DEVICE_WIREGUARD_GET_PRIVATE (object);
+	const NMPropertiesInfo property_info[] = {
+		{ NM_DEVICE_WIREGUARD_PRIVATE_KEY, &priv->private_key },
+		{ NM_DEVICE_WIREGUARD_PUBLIC_KEY,  &priv->public_key },
+		{ NM_DEVICE_WIREGUARD_LISTEN_PORT, &priv->listen_port },
+		{ NM_DEVICE_WIREGUARD_FWMARK,      &priv->fwmark },
+		{ NULL }
+	};
+
+	NM_OBJECT_CLASS (nm_device_wireguard_parent_class)->init_dbus (object);
+
+	_nm_object_register_properties (object,
+	                                NM_DBUS_INTERFACE_DEVICE_WIREGUARD,
+	                                property_info);
+}
+
+static void
+finalize (GObject *object)
```

Comment:
You don't need to implement finalize if you just chain it to the parent class.

Reply:
I see - thanks!